### PR TITLE
[RFC] Custom Scalar Specification URLs

### PR DESCRIPTION
**Pull request by eapache**
This PR contains the specification changes needed for https://github.com/graphql/graphql-spec/issues/635.

### CODE REVIEW

**Comment by sungam3r** [comment]
Relates to #633

File: spec/Section 4 -- Introspection.md
Comment (sungam3r) [inline spec/Section 4 -- Introspection.md]: It correlates to some extent with #300 .
If this extension of the introspection schema will be added, it is logical to expect that the proposed in #300 extension 
 should be also accepted. Now there is a second server directive besides `@deprecated`, which is provided through introspection as a special property. I have nothing against this, but I believe that we need to provide a way to get user directives through introspection too. It will be a consistent step.

Comment (michaelstaib) [reply spec/Section 4 -- Introspection.md]: @sungam3r there are some experiments on introspection extensions. Technically you can already today extend introspection but this alway comes with the danger that you violate a future specification. 

@IvanGoncharov was experimenting on this one with GraphQL-js. At the moment you can put metadata on almost everything there. We do almost the same with hot chocolate. The next thing would be to allow to have something like extensions on introspection so that there is a point where you can cleanly extend the introspection types.

Doing this through directives brings some challenges since directives, specifically their arguments are input types that could bear challenges to automatically extend the introspection from those.

@IvanGoncharov maybe we should start a new discussion on that?

File: spec/Section 3 -- Type System.md
Comment (andimarek) [inline spec/Section 3 -- Type System.md]: This is a bit "misleading" example because RFC3339 is normally used to describe date and time. 
How about `scalar DateTime @specified(by: "https://example.com/myDateTime")` ?

File: spec/Section 4 -- Introspection.md
Comment (andimarek) [inline spec/Section 4 -- Introspection.md]: typo: `should use the`

File: spec/Section 3 -- Type System.md
Comment (andimarek) [reply spec/Section 3 -- Type System.md]: Using a non RFC link would also make it more clear that it is totally fine to come up with your scalar spec.

**Comment by andimarek** [comment]
great work @eapache

File: spec/Section 4 -- Introspection.md
Comment (sungam3r) [reply spec/Section 4 -- Introspection.md]: I have a separate graphql-dotnet branch in which directives are already returned via introspection. We use this forked preview version in production.

> this alway comes with the danger that you violate a future specification.

I agree. It’s not scary for us. We need new features now, and not when they (possibly) appear in the specification.

File: spec/Section 3 -- Type System.md
Comment (michaelstaib) [reply spec/Section 3 -- Type System.md]: I think we all agreed to have a specific spec for datetime since it is based on the rfc but more narrow in its implementation. There should be a specific document outlining the rules for this one.

File: spec/Section 4 -- Introspection.md
Comment (michaelstaib) [reply spec/Section 4 -- Introspection.md]: @sungam3r i did mean that more from them perspective. Hc allows extending the introspection already for a couple of versions. But the user that does that has to know that he/she might violate a future spec. We just have a warning in the docs that you can do it but may run into an issue in the future.

File: spec/Section 3 -- Type System.md
Comment (eapache) [reply spec/Section 3 -- Type System.md]: Yeah, I might just take this out as an example until we have something more useful to link to.

Comment (eapache) [reply spec/Section 3 -- Type System.md]: Actually, if I take it out it gets even more confusing because the other example to talk about in the paragraph above is URL and then it gets really hard to parse whether I'm talking about the `URL` scalar as an example, or the URL that you provide in the example to link to the specification.

I'll rename this to `DateTime` for now, with the expectation that when Andi's restricted datetime scalar format doc gets published we'll change the link to that.

File: spec/Section 3 -- Type System.md
Comment (leebyron) [inline spec/Section 3 -- Type System.md]: I think throughout we can replace `RFC3986-compliant URI` with just `URL`

`RFC3986-compliant URI` could be understood to include an ISBN for an academic paper or something like that, and the goal here is that a person can load this URL in a browser and go read the spec. `URL` is well understood with a single definition, so there's no need to qualify it further with which IETF RFC you're referring to.

File: spec/Section 4 -- Introspection.md
Comment (m14t) [inline spec/Section 4 -- Introspection.md]: Nitpick: Currently in my draft for the reference implementation https://github.com/graphql/graphql-js/pull/2276, if you don't add the `@specified` directive to a scalar, `specifiedBy` will be `null`.

If that's the correct implementation, then perhaps this wording could be updated to something along the lines of:
```suggestion
  # should be a string or null for SCALAR only, must be null for the others
```

Comment (eapache) [reply spec/Section 4 -- Introspection.md]: Hmm, it's a good point but I think I'd prefer just replacing `should` with `can`.

File: spec/Section 3 -- Type System.md
Comment (eapache) [inline spec/Section 3 -- Type System.md]: @leebyron I changed the capitalization here from `Url` to `URL` because `Uuid` looked particularly gross, and the built-in is already `ID` not `Id`. It's a minor editorial thing but it's not really related to this RFC so I wanted to make sure it was ok.

File: spec/Section 4 -- Introspection.md
Comment (IvanGoncharov) [inline spec/Section 4 -- Introspection.md]: We don't have a `URL` type so it should be `String`.

**Comment by IvanGoncharov** [comment]
@eapache bikeshedding: How about changing `@specified(by: 'http://example.com')` to `@specifiedBy(url: 'http://example.com')`
Pros:
* Fewer chances to conflict with existing and previous directives
* It's easier to figure out that you pass/get URL from name, e.g. inside introspectoon we would have `specifiedByURL` instead of `specifiedBy`

**Comment by sungam3r** [comment]
+1 for `@specifiedBy(url: 'http://example.com')`
-1 for `specifiedByURL`

This is a personal feelings, `specifiedByURL` causes some associations with the Hungarian notation, which has its drawbacks.

File: spec/Section 4 -- Introspection.md
Comment (eapache) [reply spec/Section 4 -- Introspection.md]: maybe `may return a String (in the form of a URL) or {null}.` ?

**Comment by eapache** [comment]
I'm weakly in favour of `@specifiedBy(url:)` instead of `@specified(by:)`.

File: spec/Section 3 -- Type System.md
Comment (m14t) [inline spec/Section 3 -- Type System.md]: While implementing this in https://github.com/graphql/graphql-js/pull/2276 @IvanGoncharov  provided the following [comment](https://github.com/graphql/graphql-js/pull/2276#discussion_r358119871):
> If at some point we decide to add `UUID` scalar into this library it would be a confusing example.
> It's better to use some abstract scalar like `Foo` and `https://example.com/foo_spec`.

**Question:** Do we want to make that same change here as well?  If so this likely applies to `URL` as well, and effects a few locations in this PR.

File: spec/Section 3 -- Type System.md
Comment (m14t) [inline spec/Section 3 -- Type System.md]: I believe this project prefers American spelling.
Reasoning: `color` is in this project 74 times, compared to `colour` which is never used.

```suggestion
to provide a URL for specifying the behavior of custom
```

**Comment by IvanGoncharov** [comment]
Reviewing https://github.com/graphql/graphql-js/pull/2276 I saw another argument for naming argument `url` instead of `by`, since `by` by itself looks weird in code:
https://github.com/graphql/graphql-js/pull/2276/files#diff-71ba52e9c625f826d2b0df2963c8633aR328

File: spec/Section 3 -- Type System.md
Comment (leebyron) [reply spec/Section 3 -- Type System.md]: Seems totally fine to me

File: spec/Section 3 -- Type System.md
Comment (leebyron) [inline spec/Section 3 -- Type System.md]: This could fit in one line

File: spec/Section 3 -- Type System.md
Comment (leebyron) [inline spec/Section 3 -- Type System.md]: It would be great to elaborate a bit here about how we expect client tooling to use this to provide additional behavior, and maybe even a note about how two GraphQL services with different named scalars specified by the same url should be imbued with the same behavior and conversely two services might have a scalar by the same name but specified with different urls.

File: spec/Section 3 -- Type System.md
Comment (leebyron) [inline spec/Section 3 -- Type System.md]:
```suggestion
a GraphQL system providing a `UUID` scalar might link to [RFC 4122](https://tools.ietf.org/html/rfc4122),
```

File: spec/Section 3 -- Type System.md
Comment (leebyron) [inline spec/Section 3 -- Type System.md]: > If a specification URL is present, systems must conform to the described rules.

I don't think this is possible as "must". If you publish a URL tomorrow with a spec and define a custom scalar with it, there's no way my GraphQL system build today will know how to conform to its rules.

Maybe this can be something like "If a specification URL is present, GraphQL services and client tools should conform to its described rules if they are aware of that URL."

File: spec/Section 4 -- Introspection.md
Comment (leebyron) [inline spec/Section 4 -- Introspection.md]:
```suggestion
A GraphQL type designer should define `specifiedBy` for custom scalars as
```

File: spec/Section 3 -- Type System.md
Comment (eapache) [reply spec/Section 3 -- Type System.md]: Point taken that it could become confusing, but it's not confusing yet, and there is definitely value in having a concrete example that lines up with use cases a lot of actual developers will run into. I'm planning to leave this as is, and we can change it if becomes an issue later.

File: spec/Section 3 -- Type System.md
Comment (eapache) [reply spec/Section 3 -- Type System.md]: Whoops, my mistake, I knew that and just forgot.

File: spec/Section 3 -- Type System.md
Comment (eapache) [reply spec/Section 3 -- Type System.md]: Hmm, that's a good idea. Since I believe this is basically ready to go as-is (once the implementation lands, which should be soon?) I might put that content into a second PR to avoid holding this one up over additional wording quibbles.

**Comment by andimarek** [comment]
This PR introduces another built-in directives next to `deprecated`, `skip` and `include`. 

The question are: (which also affect the other built-ins):
- Should they be returned by Introspection? (Probably yes)
- Should they be included in SDL when printed? (Probably no)
(The question above are addressed by https://github.com/graphql/graphql-spec/pull/633/files)

Additionally: are you allowed to redefine `specifiedBy`? (Same question for the other built-ins) GraphQL.js currently allows redefining `deprecated`, `skip` and `include`. If you allow  redefining them the server implementations need to distinguish between the built-ins `specifiedBy` and the custom `specifiedBy`.

I lean towards not allowing to redefine them, but this would make this new directive a breaking change for all schemas which already define a `specifiedBy` directive.

**Comment by IvanGoncharov** [comment]
@andimarek Realistically speaking, how many API have `specifiedBy` directive?
I would say zero, with the exception of implementations motivated by this PR.

I don't think we should increase scope to resolve purely theoretical issues.
My opinion `specifiedBy` should be treated exactly the same as `@if`, `@skip` or `@deprecated`.

**Comment by sungam3r** [comment]
I agree with @IvanGoncharov, although such judgments always give a special taste to statements about the priority of backward compatibility.

**Comment by andimarek** [comment]
@IvanGoncharov I agree with not increasing the scope of this change. But I think we should follow up with some additional clarification how to handle custom directives called `specifiedBy`, `deprecated` etc.

Sure it will only affect a very small minority of schemas, but we should nevertheless care.

It could be as simple as saying that custom directives named `specifiedBy` etc. are not allowed and pointing out it in the release notes for the next spec release that it is a theoretical breaking change. I think this would be sufficient.

**Comment by andimarek** [comment]
@leebyron @IvanGoncharov are there any issues for this PR that need to be addressed before it is ready?

**Comment by leebyron** [comment]
Will take a careful read-over before merging

**Comment by leebyron** [comment]
Notes from meeting:

URL example is a good one, but it would be useful to have an example showing a vendor example.

**Comment by eapache** [comment]
Just realized this was still open. Is there anything left to do besides hitting the button on this one?

**Comment by leebyron** [comment]
Sorry for the flurry of editorial edits - I'm working on ratifying the next edition of the spec and working through merging accepted proposals in the process.

This one had more significant editorial edits than most so I'm going to leave it open for a little while in case there is any feedback or issue with those changes

File: spec/Section 3 -- Type System.md
Comment (eapache) [inline spec/Section 3 -- Type System.md:344]: I agree with moving custom scalars to their own section, but I find the resulting paragraph here a bit confusing as it refers to two separate things "below" in a way that's somewhat ambiguous.

File: spec/Section 3 -- Type System.md
Comment (eapache) [inline spec/Section 3 -- Type System.md:375]: You editorialized out a `the` here to make the sentence consistent, but grammatically I think they _both_ need articles? So "via the `specifiedBy` introspection field or the `@specifiedBy` directive."

File: spec/Section 3 -- Type System.md
Comment (eapache) [inline spec/Section 3 -- Type System.md:391]: semi-colon seems incorrect here, I think a comma is fine?

Comment (eapache) [reply spec/Section 3 -- Type System.md:391]: Or, drop the word "as" to make the second clause a complete sentence.

File: spec/Section 3 -- Type System.md
Comment (leebyron) [reply spec/Section 3 -- Type System.md:344]: Good point. Let me think about how to clarify

File: spec/Section 3 -- Type System.md
Comment (leebyron) [reply spec/Section 3 -- Type System.md:375]: That was a mistake on my part, thanks for the careful read

File: spec/Section 3 -- Type System.md
Comment (leebyron) [reply spec/Section 3 -- Type System.md:391]: Great catch, thank you!

**Review by benjie — APPROVED** [review]
Looks good to me; other than the issues @eapache has already highlit and @leebyron is addressing.

**Pull request merged**
